### DISCUSSION TITLE
Official homebrew support (3ds, Wii U, PS3, Xbox 360)

### DISCUSSION THREAD

**Discussion by FlooferLand** [godotengine/godot-proposals#10698] [Platforms]
I and [quite a few people _(upvotes)_](https://github.com/godotengine/godot-docs-user-notes/discussions/28#discussioncomment-9693079) wish there were official export templates for older vintage consoles like the Nintendo 3ds and the Wii U.

These consoles have been long abandoned, but there are still thousands of people trying to make games for them because the consoles aged like fine wine, and services such as [Pretendo](https://pretendo.network) have popped up the past few years letting you even bring back multiplayer services for those consoles _(I'm willing to make an addon to bring Pretendo if this engine ever gets 3ds/Wii U support)_.

Currently, your only option is using a decade old version of Unity, or writing your own engine from scratch using C.
There are projects like [Homebrodot](https://github.com/Homebrodot) but they're not actively maintained and tend to be very out of date _(requiring ancient versions of Godot like 2.1)_

All the mentioned consoles support OpenGL ES 1.0 or 2.0, so they should be doable with the compatibility renderer. _(I don't expect to be running Vulkan on the Nintendo 3Ds)_
There would need to be custom implementations for handling audio and other low-end stuff, but if there is anyone on here that has developed for these consoles it shouldn't be that difficult.

I would love it if I was able to just sit down and make Wii U games at the comfort of my Godot 4.3 or even port existing games to it.
This would open up homebrew development to tons of people.

**Comment by AThousandShips** [comment]
Please see [this](https://godotengine.org/article/about-official-console-ports/) article about exactly this for the answers to this (or rather, an answer to the general topic which applies to homebrew as well with the legal liability etc.), and homebrew doesn't necessarily free one from the licensing issues, even reverse-engineering is a legal minefield, and if it isn't clean-room reverse-engineered it's still under NDAs etc.

**Reply by AThousandShips** [reply]
And beyond the points raised there one of my personal concerns with supporting any kind of homebrew things for console is the risk of causing harm to developers and end users who don't consider the risks sufficiently or don't get the whole picture

Nintendo especially are known to be very aggressive when it comes to users using various homebrew and unofficial hardware or builds, with users being banned for using their accounts on such devices or on jailbroken devices, losing any games they've bought in the process, by supporting things that aren't supported by Nintendo etc. we expose users to that risk and I feel that's something we should be careful about. If people don't read carefully how to avoid issues when using these homebrew systems etc. and end up getting banned that's bad, and we'd be at least partially responsible in my opinion for that risk

**Reply by FlooferLand** [reply]
I know Nintendo is aggressive about jailbreaking the Switch and emulation and all.
They don't really seem to care about homebrew for old consoles or even reverse engineering attempts for them.
The Pretendo network has been huge for the last like 5 years and the entire project is a 3ds / Wii U reverse engineering attempt and they're safe. There shouldn't be any emulation required for actually building out games, so Godot should be safe on that end too. There are unofficial projects out there that have provided ways to make games for these consoles for years, they're just C++ libraries.

As another note I wonder if W4 could contact Nintendo about homebrew if they ever strike a deal for Nintendo Switch exports.

**Reply by AThousandShips** [reply]
> As another note I wonder if W4 could contact Nintendo about homebrew if they ever strike a deal for Nintendo Switch exports.

The same things for licensing and NDAs would apply

Also small scale projects doing things and Godot doing things officially is very different when it comes to what Nintendo and others might consider a "threat" and something worthy of targeting with a takedown notice or a lawsuit

But regardless homebrew is largely legally complicated and far from simple

**Comment by bruvzg** [comment]
> All the mentioned consoles support OpenGL ES 1.0 or 2.0, so they should be doable with the compatibility renderer.

It's not, compatibility renderer require OpenGL ES 3.0.

**Comment by Meorge** [comment]
While I personally love the idea of being able to take advantage of consoles' unique features with my own games built in Godot, I don't think it would make sense for them to be officially supported, mostly for the reasons AThousandShips provided.

As a Nintendo fan I know how litigious they can get, and I wouldn't want Godot to come under fire for it. Even if Nintendo (or other companies) didn't take any *direct* action against Godot, I could see it hampering possibilities of better console support in the future.

Additionally, making support for it "official" would require developers who are knowledgeable in these consoles to start working on it, which would probably be the same people as have worked on things like Hombrodot. If those people stopped working on their own unofficial support for the consoles, I'm not sure that official support could be maintained that lives up to the standards we would expect.

Ultimately, I think the best approach for this would be to have unofficial ports be worked on by independent developers. That would keep Godot itself away from drawing the ire of companies who don't want their consoles developed for in this way (as much as it can, being a FOSS product), while also allowing the devs who are interested in it to more quickly make progress on it.
